bin/inx-chronicle/src/stardust_inx/mod.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod cone_stream;
+mod config;
+mod error;
+mod milestone_stream;
+mod syncer;
+
+use async_trait::async_trait;
+use chronicle::{
+    db::MongoDb,
+    runtime::{Actor, ActorContext, ActorError, ConfigureActor, HandleEvent, Report},
+};
+pub use config::InxConfig;
+pub use error::InxError;
+use inx::{client::InxClient, proto::NoParams, tonic::Channel, NodeStatus};
+pub use milestone_stream::MilestoneStream;
+
+use self::{
+    config::SyncKind,
+    syncer::{SyncNext, Syncer},
+};
+
+pub struct Inx {
+    db: MongoDb,
+    config: InxConfig,
+}
+
+impl Inx {
+    /// Creates an [`InxClient`] by connecting to the endpoint specified in `inx_config`.
+    pub fn new(db: MongoDb, inx_config: InxConfig) -> Self {
+        Self { db, config: inx_config }
+    }
+
+    pub async fn connect(inx_config: &InxConfig) -> Result<InxClient<Channel>, InxError> {
+        let url = url::Url::parse(&inx_config.connect_url)?;
+
+        if url.scheme() != "http" {
+            return Err(InxError::InvalidAddress(inx_config.connect_url.clone()));
+        }
+
+        InxClient::connect(inx_config.connect_url.clone())
+            .await
+            .map_err(InxError::ConnectionError)
+    }
+}
+
+#[async_trait]
+impl Actor for Inx {
+    type State = ();
+    type Error = InxError;
+
+    async fn init(&mut self, cx: &mut ActorContext<Self>) -> Result<Self::State, Self::Error> {
+        log::info!("Connecting to INX at bind address `{}`.", &self.config.connect_url);
+        let mut inx_client = Self::connect(&self.config).await?;
+        log::info!("Connected to INX.");
+
+        // Request the node status so we can get the pruning index and latest confirmed milestone
+        let node_status = NodeStatus::try_from(inx_client.read_node_status(NoParams {}).await?.into_inner())
+            .map_err(InxError::InxTypeConversion)?;
+        let latest_ms = node_status.confirmed_milestone.milestone_info.milestone_index;
+        let configured_start = match self.config.sync_kind {
+            SyncKind::Max(ms) => latest_ms - ms,
+            SyncKind::From(ms) => ms,
+        };
```

Comment:
As far as I can tell, we never really throw away milestones right now. Because of this I think this is kind of confusing as the dataset can still go beyond the size.

Maybe we leave this out for now and revisit it once we do the selective permanode functionality?

Reply:
This has nothing to do with selection. The config allows either choosing a specific starting milestone or the milestone N ago.

Reply:
Do you think a `starting_milestone: Option<MilestoneIndex>` would be more descriptive? The current config is tricky to understand, especially without comments.